httputil/client.go

```diff
@@ -21,16 +21,130 @@ package httputil
 
 import (
 	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
 	"net/http"
 	"net/url"
+	"path/filepath"
 	"time"
+
+	"github.com/snapcore/snapd/logger"
 )
 
+// CertData contains the raw data of a certificate and the origin of
+// the cert, this is usually a file path on disk and is just used
+// for error reporting.
+type CertData struct {
+	Raw    []byte
+	Origin string
+}
+
+// ExtraSSLCerts is an interface that provides a way to add extra
+// SSL certificates to the httputil.Client
+type ExtraSSLCerts interface {
+	Certs() ([]*CertData, error)
```

Comment:
I slightly dislike the fact that we're loading all certs into the memory and then passing them to the caller. This isn't much of a problem for a single PEM cert, which has to be read anyway. However, should there be multiple certificates in a directory, we'll end up loading all of them into the memory. The PEM files tend to be quite big even with a single certificate inside.

Would an interface like this make sense?
```
type ExtraSSLCertLoader interface {
	LoadInto(pool *x509.CertPool) error
}
```

Then `ExtraSSLCertsFromDir` would iterate dir entries like it does now and load the cert right away (much like x509.SystemCertPool() does):

```
func (e *ExtraSSLCertsFromDir) LoadInto(pool *x509.CertPool) error {
	if pool == nil {
		return fmt.Errorf("internal error: certificate pool not provided")
	}
	extraCertFiles, err := filepath.Glob(filepath.Join(e.Dir, "*.pem"))
	if err != nil {
		return err
	}
	extraCerts := make([]*CertData, 0, len(extraCertFiles))
	for _, p := range extraCertFiles {
		certPEM, err := ioutil.ReadFile(p)
		if err != nil {
			return fmt.Errorf("cannot read extra certificate: %v", err)
		}
		if ok := pool.AppendCertsFromPEM(certPEM); !ok {
			return fmt.Errorf("cannot append store ssl certificate: %v", p)
		}
	}
	return nil
}
```

Reply:
We don't expect many certificates to be set this way, so I would keep the naive interface for now. We can revisit later.